### ISSUE TITLE
[API] Missing tensors() method in Model trait causes incomplete abstraction

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#332]
# [API] Missing `tensors()` method in Model trait causes incomplete abstraction

## Problem Description

The `Model` trait in `/home/steven/code/Rust/BitNet-rs/crates/bitnet-models/src/bitnet.rs` is missing a `tensors()` method that is being called in the model loader. The loader code at line 127 in `/home/steven/code/Rust/BitNet-rs/crates/bitnet-inference/src/loader.rs` attempts to call `model.tensors().len()` to count loaded tensors, but this method is not defined in the trait interface.

**Current Code Location:**
- **File:** `crates/bitnet-inference/src/loader.rs`
- **Line:** 127
- **Code:** `let tensors_loaded = model.tensors().len();`

## Environment

- **OS:** Linux 6.6.87.2-microsoft-standard-WSL2
- **Rust Version:** 1.90.0+ (MSRV)
- **Feature Flags:** `--no-default-features --features cpu`
- **Build Configuration:** Default development profile

### Current Model Trait Definition
The `Model` trait currently defines:
```rust
pub trait Model: Send + Sync {
    fn config(&self) -> &BitNetConfig;
    fn forward(&self, input: &ConcreteTensor, cache: &mut dyn std::any::Any) -> Result<ConcreteTensor>;
    fn embed(&self, tokens: &[u32]) -> Result<ConcreteTensor>;
    fn logits(&self, hidden: &ConcreteTensor) -> Result<ConcreteTensor>;
}
```

### Missing Method
The trait lacks a `tensors()` method that should return information about the loaded tensors, which is essential for:
1. **Loader Metadata**: Counting tensors for diagnostic/logging purposes
2. **Model Introspection**: Debugging model structure and weights
3. **Memory Management**: Understanding tensor allocation patterns
4. **Validation**: Ensuring all expected tensors are loaded

### Impact Assessment
- **Severity:** Medium - Affects model introspection and debugging capabilities
- **Scope:** All model implementations must provide tensor enumeration
- **Technical Debt:** Incomplete trait abstraction reduces API consistency
- **Maintainability:** Makes it difficult to uniformly inspect model state across implementations

## Reproduction Steps

1. Navigate to the model loader implementation
2. Examine line 127 in `crates/bitnet-inference/src/loader.rs`
3. Observe the call to `model.tensors().len()`
4. Check the `Model` trait definition in `crates/bitnet-models/src/bitnet.rs`
5. Note that no `tensors()` method is defined in the trait

**Expected Result:** The trait should define a `tensors()` method
**Actual Result:** The method is missing from the trait interface

### Primary Implementation
Add a `tensors()` method to the `Model` trait that returns tensor metadata suitable for introspection:

```rust
pub trait Model: Send + Sync {
    fn config(&self) -> &BitNetConfig;
    fn forward(&self, input: &ConcreteTensor, cache: &mut dyn std::any::Any) -> Result<ConcreteTensor>;
    fn embed(&self, tokens: &[u32]) -> Result<ConcreteTensor>;
    fn logits(&self, hidden: &ConcreteTensor) -> Result<ConcreteTensor>;

    /// Return tensor metadata for introspection and debugging
    /// Returns a collection that can be used for counting and inspection
    fn tensors(&self) -> Box<dyn Iterator<Item = (&str, &dyn std::any::Any)> + '_>;
}
```

### BitNetModel Implementation
```rust
impl Model for BitNetModel {
    // ... existing methods ...

    fn tensors(&self) -> Box<dyn Iterator<Item = (&str, &dyn std::any::Any)> + '_> {
        Box::new(self.tensors.iter().map(|(name, tensor)| (name.as_str(), tensor as &dyn std::any::Any)))
    }
}
```

### Alternative Approach (Simpler)
If full introspection is not needed immediately, provide a minimal implementation:

```rust
pub trait Model: Send + Sync {
    // ... existing methods ...

    /// Return the number of loaded tensors
    fn tensor_count(&self) -> usize;

    /// Return tensor names for debugging
    fn tensor_names(&self) -> Vec<String>;
}
```

### Phase 1: Trait Extension
- [ ] Add `tensors()` method to `Model` trait in `bitnet-models/src/bitnet.rs`
- [ ] Choose appropriate return type (iterator vs. collection vs. count)
- [ ] Update trait documentation with usage examples

### Phase 2: Implementation
- [ ] Implement `tensors()` method in `BitNetModel`
- [ ] Implement `tensors()` method in `MockBitNetModel`
- [ ] Ensure all trait implementors provide the method

### Phase 3: Integration
- [ ] Update loader code to use the new trait method
- [ ] Verify loader metadata correctly reflects tensor counts
- [ ] Add logging/tracing for tensor enumeration

### Phase 4: Testing
- [ ] Add unit tests for tensor enumeration
- [ ] Test with various model formats (GGUF, SafeTensors)
- [ ] Validate tensor counting accuracy across model types
- [ ] Test edge cases (empty models, corrupted tensors)

### Unit Tests
```rust
#[test]
fn test_model_tensor_count() {
    let model = create_test_model();
    let tensor_count = model.tensors().count(); // or model.tensor_count()
    assert!(tensor_count > 0, "Model should have loaded tensors");
}

#[test]
fn test_tensor_introspection() {
    let model = load_test_gguf_model();
    let tensors: Vec<_> = model.tensors().collect();
    assert!(tensors.iter().any(|(name, _)| name.contains("embed")));
    assert!(tensors.iter().any(|(name, _)| name.contains("output")));
}
```

### Integration Tests
- Load real GGUF models and verify tensor enumeration
- Test loader metadata accuracy with tensor counting
- Cross-validate with C++ reference implementation counts

## Related Issues

This issue is part of a broader effort to eliminate stub implementations and improve API completeness:

- **Related:** Issue #stub-model-config (Model configuration validation)
- **Blocks:** Model introspection and debugging workflows
- **Enables:** Better model loading diagnostics and validation

## Acceptance Criteria

- [ ] `Model` trait defines a `tensors()` or equivalent method
- [ ] All trait implementations provide working tensor enumeration
- [ ] Loader code successfully uses the trait method instead of undefined behavior
- [ ] Tensor counting is accurate and matches loaded model state
- [ ] Method is well-documented with usage examples
- [ ] Unit and integration tests validate the implementation
- [ ] No breaking changes to existing API consumers

### Memory Efficiency
- Use iterators where possible to avoid collecting all tensor metadata at once
- Consider lazy evaluation for large models with many tensors

### Type Safety
- Ensure tensor references remain valid for the iterator lifetime
- Use appropriate trait bounds to prevent unsafe access

### Cross-Platform Compatibility
- Verify implementation works across CPU/GPU backends
- Test with different quantization formats (I2_S, TL1, TL2)

### Performance Impact
- Tensor enumeration should be O(n) where n is the number of tensors
- Avoid expensive operations during enumeration (tensor copying, etc.)

## Documentation Updates Required

- [ ] Update trait documentation with new method
- [ ] Add examples of tensor introspection workflows
- [ ] Document relationship between loader metadata and tensor enumeration
- [ ] Update architecture documentation to reflect improved model abstraction
